### DISCUSSION TITLE
how to combine query parameters using depends() with model column/field filtering

### DISCUSSION THREAD

**Discussion by Allena101** [fastapi/fastapi#11116] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class KWh(Base):
    __tablename__ = "eKWh"
    id = Column(Integer, primary_key=True, nullable=False)
    dt = Column(TIMESTAMP(timezone=True), nullable=False)
    KWhT1 = Column(Float, nullable=False)
    KWhT2 = Column(Float, nullable=False)
    KWhT3 = Column(Float, nullable=False)
    KWhT4 = Column(Float, nullable=False)

class SeriesNum(str, Enum):
    kWh_default = ""
    kwh1 = "KWhT1"
    kwh2 = "KWhT2"
    kwh3 = "KWhT3"
    kwh4 = "KWhT4"


class CommonQueryParams(BaseModel):
    seriesNum: SeriesNum = Field(
        Query("", description="Please pick ONLY one of the 4 available KWh series")
    )
    city_name: CityEnum = Field(
        Query("", description="Please pick ONLY one of the 2 available cities")
    )
    google: List[GoogleEnumList] = Field(
        Query([], description="Please pick any of the 3 available google searches")
    )
    start_dt: str = Field(
        Query(
            "2020-06-18 10",
            description="Start date for timeSeries",
            example="2021-11-21 14",
        )
    )
    end_dt: str = Field(
        Query(
            "2024-01-12 12",
            description="End date for timeSeries",
            example="2022-11-21 14",
        )
    )
    pred_horizon: int = Field(
        Query(
            7,  # 7 days
            gt=0,
            lt=32,
            description="How many timeSteps the model will forecast",
        )
    )
    show_image: bool = Field(
        Query(False, description="Returns image using fastAPI FileResponse")
    )
    stockmarket: bool = Field(
        Query(False, description="Use omxs30 timeSeries data as covariate")
    )

    @validator("seriesNum")
    def validate_series_num(cls, v):
        if not v:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="seriesNum is required",
            )
        return v

    @validator("start_dt")
    def validate_start_dt(cls, v):
        """Validates that start_dt is in the format yyyy-mm-dd hh."""
        try:
            dt = datetime.datetime.strptime(v, "%Y-%m-%d %H")
        except ValueError as exc:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"start_dt {v} has to be in the format of YYYY-MM-DD HH. For example: 2022-06-18 10",
            ) from exc

        if dt < datetime.datetime(2020, 6, 18) or dt > datetime.datetime(2024, 1, 12):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"start_dt cannot be earlier than 2020-06-18 or later than 2024-01-12. Received: {dt}",
            )

        return v

    @validator("end_dt")
    def validate_end_dt(cls, v):
        """Validates that end_dt is in the format yyyy-mm-dd hh."""
        try:
            dt = datetime.datetime.strptime(v, "%Y-%m-%d %H")
        except ValueError as exc:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"end_dt {v} has to be in the format of YYYY-MM-DD HH \n For example: 2022-06-18 10",
            ) from exc
        if dt < datetime.datetime(2020, 6, 18) or dt > datetime.datetime(2024, 1, 12):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"end_dt cannot be earlier than 2020-06-18 or later than 2024-01-12. Received: {dt}",
            )

        return v

    @validator("end_dt")
    def validate_dt_order(cls, v, values):
        """Validates that end_dt is after start_dt."""
        start_dt = values.get("start_dt")
        end_dt = datetime.datetime.strptime(v, "%Y-%m-%d %H")
        start_dt = datetime.datetime.strptime(start_dt, "%Y-%m-%d %H")
        if end_dt <= start_dt:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"end_dt: {end_dt} does not come after start_dt: {start_dt}",
            )
        return v



@app.get("/test")
def test(
    db: db_dependency,
    query_params: CommonQueryParams = Depends()
):

    kwh = (
        db.query(*[KWh.dt], seriesNumDict[query_params.seriesNum.value])
        .filter(KWh.dt >= start_dt_with_tz)
        .filter(KWh.dt <= end_dt_with_tz)
        .all()
    )

   if query_params.city_name:
        if query_params.city_name.value not in city_name_list:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"City name: {query_params.city_name.value} is not in the database. \nCurrently there is only data for Stockholm and Karlstad",
            )

        if query_params.city_name == "Stockholm":
            weather_list = ["Sthlm_temp", "Sthlm_precip"]
        if query_params.city_name == "Karlstad":
            weather_list = ["Ksd_temp", "Ksd_precip"]

        weatherList = []
        for i in weather_list:
            weatherList.append(weather_dict[i])

        weather_data = (
            db.query(Weather.dt, *weatherList)
            .filter(Weather.dt >= start_dt_with_tz)
            .filter(Weather.dt <= end_dt_with_tz)
            .all()
        )




    # lots of processing

     if show_image:
        return {"preds_KWh": preds_KWh, "base64_data": base64_data}
```

### Description

Hello everyone! 👋

My question/issue is not an error, its more of confusion surrounding query parameters and model filtering. I did not include all the tables - only one in the code. But i think (perhaps i am wrong about this) it will highlight my issue.

So the issue is that in all the guides (and the documentation) pydantic models are used to enforce types for database tables. And this makes sense and is very clear. What is not as clear, and i have found several people asking about this (but not my exact issue) - is how to filter models and also have requirements on the query parameters.

I havent even seen an example where more than one pydantic model is used in an endpoint. In the code i am working on. i want to query from 1-4 tables and still have restrictions and validation on the query params.

My main issue is that i have gotten it to work (with many problems along the way), but it feels like i am doing it wrong, especially sicne i dont see anyone doing it this way.

I think it should be possible to supply the endpoint with one model for the query params (e.g. start_dt , end_dt) and then one model for each table (plus a response model that would be either an image or an array). 

One major issue that i also run into is that not all columns will always be queried, but in all examples/guides - all the columns are usually queried each time. And i am not talking about when you create a new pydantic model that excludes one column (e.g. password) that you dont want to return from the database. For example, in my case, the user might only want to query from 2 out of 4 tables from the database (only 1 out of the 4 tables are required - KWh). 

The only solution i can think of that could work is to have the , for example in the KWh table, each column be optional , but also have a requirement that one of them are chosen. I manage to fulfill this requirement by creating the Enum class and in the swagger docs, the user can only choose one of the four columns. 

I have also seen examples where a pydantic model is created that receives a join query from several tables. But this would not work in my case since i need to do processing on the data after the query so i can join them later in a dataFrame.  

to summaries - I want to provide the endpoint with 4 pydantic models (only one of them are required) and one model for the query parameters. 

Is that the proper way to do this? 

I apologize if this question is not clear. I might be thinking wrong when it comes to how to structure the whole thing. 

### Operating System

Windows

### FastAPI Version

0.108.0

### Pydantic Version

2.5.3

### Python Version

3.10

**Comment by Allena101** [comment]
I also want to add, which contributes to the feeling that i am doing something wrong, is with the examples in the doc (Multiple Models with FastAPI) one endpoint is still used per model. But in my case, i want to query 1-3 tables and then do some processing on the results. I remember reading that it was not a good idea to call your own endpoint in another endpoint, though that would be a workaround since each endpoint could then have its own model linked to it

**Comment by Allena101** [comment]
Here are the the other tables:

```py

class Weather(Base):
    __tablename__ = "weather"
    id = Column(Integer, primary_key=True, nullable=False)
    dt = Column(TIMESTAMP(timezone=False), nullable=False)
    Ksd_temp = Column(Float, nullable=False)
    Ksd_precip = Column(Float, nullable=False)
    Sthlm_temp = Column(Float, nullable=False)
    Sthlm_precip = Column(Float, nullable=False)


class gTrend(Base):
    __tablename__ = "gTrend"
    id = Column(Integer, primary_key=True, nullable=False)
    dt = Column(TIMESTAMP(timezone=False), nullable=False)
    skatt = Column(String, nullable=False)
    el = Column(String, nullable=False)
    hyra = Column(String, nullable=False)


class omxs30(Base):
    __tablename__ = "omxs30"
    id = Column(Integer, primary_key=True, nullable=False)
    dt = Column(TIMESTAMP(timezone=False), nullable=False)
    open = Column(Float, nullable=False)
    close = Column(Float, nullable=False)
```

**Reply by binbjz** [reply]
### How to Combine Query Parameters with Model Column/Field Filtering Using `Depends()`

FastAPI's `Depends()` dependency system allows you to create reusable dependencies (e.g., database sessions, authentication information, query parameter parsing, etc.) and use these dependencies across multiple path operations. Combining query parameters with model column/field filtering typically involves the following steps:

1. **Defining Query Parameter Models**: You have already completed this step with the `CommonQueryParams`, which is used to parse and validate query parameters in the request.

2. **Injecting Query Parameters with `Depends()`**: In the path operation function, you can use `Depends()` to inject an instance of `CommonQueryParams`. FastAPI will automatically parse the query parameters from the request and validate them according to your Pydantic model.

3. **Filtering Database Queries Based on Query Parameters**: Within the path operation function, you can construct database queries based on the values in the `CommonQueryParams` instance. This may involve dynamically selecting columns to query based on the user's choice and applying filtering conditions.

For example, if you want to filter `KWh` data based on the user's input of `seriesNum`, you could do it like this:

```python
from fastapi import Depends, FastAPI, HTTPException
from sqlalchemy.orm import Session
from .database import SessionLocal, engine
from . import models

app = FastAPI()

# Dependency: Obtain a database session
def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()

@app.get("/data")
async def get_data(query_params: CommonQueryParams = Depends(), db: Session = Depends(get_db)):
    # Select column based on seriesNum
    series_column = getattr(models.KWh, query_params.seriesNum, None)
    if not series_column:
        raise HTTPException(status_code=400, detail="Invalid seriesNum")

    # Construct the query
    query_result = db.query(models.KWh.dt, series_column).all()

    # ... Further processing of the query results as needed

    return query_result
```

Note that this example assumes you have set up SQLAlchemy database session management (`SessionLocal`) and ORM models. You would need to adjust this example based on your own database configuration and model structures.

By combining `Depends()` and Pydantic models, you can create powerful and flexible API endpoints capable of receiving complex query parameters while ensuring these parameters comply with your specifications. Furthermore, you can perform complex database queries and data processing logic based on these parameters.

**Reply by Allena101** [reply]
thanks alot for reading my issue and replying binbjz!

I think i am doing the depends part correctly, though i did not use the getattr method. i used a workaround instead with a dictionary. 

series_column = getattr(models.KWh, query_params.seriesNum, None) . I assume that models.KWh is the pydantic schema of the table and NOT the sqlalchemy model,  is that correct? Also models.  , that means that you would import the models from a models.py file , right?

What confuses me alot is the omission of the model in the path operation function. How come that usually the model is used as argument in path operation function, but not in your example?

So in your example, would i be able to continue filtering using the query_params as well?
Like this , but for the weather table:

start_dt = getattr(models.WeatherPyd, query_params.start_dt, None)
end_dt = getattr(models.WeatherPyd, query_params.end_dt, None)

   weather_data = (
            db.query(models.WeatherPyd.dt, city)
            .filter(Weather.dt >= start_dt)
            .filter(Weather.dt <= end_dt )
            .all()
        )

Here is the sqlalchemy model and the pydantic schema:

```py
class Weather(Base):
    __tablename__ = "weather"
    id = Column(Integer, primary_key=True, nullable=False)
    dt = Column(TIMESTAMP(timezone=False), nullable=False)
    Ksd_temp = Column(Float, nullable=False)
    Ksd_precip = Column(Float, nullable=False)
    Sthlm_temp = Column(Float, nullable=False)
    Sthlm_precip = Column(Float, nullable=False)
    
class WeatherPyd(BaseModel):
    id: int
    dt: datetime
    Ksd_temp: float
    Ksd_precip: float
    Sthlm_temp: float
    Sthlm_precip: float
```

Also if models.KWh is only referencing the sqlalchemy model and not the pydantic schema , then doenst that mean that i am not using any pydantic models?

**Reply by Allena101** [reply]
Could part of my confusion be centered around that you are more meant to use the pydantic schemas for get methods where you are altering stuff in the database? Is post method only meant to use the response_model for the ouput and then nothing for the input (in terms of pydantic validation)?

**Reply by RichieB2B** [reply]
This is the best documentation I have found for using `Depends()` to inject query parameters in an endpoint.

One downside of using `Depends()` for query parameters is that `CommonQueryParams.model_fields_set` contains all of the class members. This means you cannot make a distinction between a member that was part of the actual request or simply a default value of the class initialization. This does work when using the [documented validation feature](https://fastapi.tiangolo.com/tutorial/query-params-str-validations/):
```python
async def get_data(query_params: Annotated[CommonQueryParams, Query()], db: Session = Depends(get_db)):
```
Now `CommonQueryParams.model_fields_set` only contains the parameters that were part of the actual request.

**Comment by jgould22** [comment]
From a quick read it sounds like you might be trying to do a bit too much with one endpoint or perhaps should be using a POST instead of a GET.

**Reply by Allena101** [reply]
thanks for replying jguld22 ! 

Yes, but i dont know how you would solve this if you are not supposed to use other endpoints in endpoint. 
From my understanding , Get is correct since i am not altering anything in the database, i am only retrieving data.

**Reply by jgould22** [reply]
Some things to think about 

> So the issue is that in all the guides (and the documentation) pydantic models are used to enforce types for database tables.

This is not really correct, pydantic is there to validate request and response data (among other things ). It does not really have anything to do with the database. It looks like you are using SQLAlchemy to interact with your database.

> What is not as clear, and i have found several people asking about this (but not my exact issue) - is how to filter models and also have requirements on the query parameters. 

It is not super clear to me what you mean here, can you expand on what you mean by "filter models"

> I remember reading that it was not a good idea to call your own endpoint in another endpoint, though that would be a workaround since each endpoint could then have its own model linked to it

If you find yourself needing code that is in another end point then it is time to extract that code and put it somewhere else, then call it from both endpoints.

I think you are thinking that the REST interface and the data base are strongly coupled and one directly reflects the other. This is sometimes true but rarely.

**Comment by YuriiMotov** [comment]
Let's close this for now to keep the discussions clean and organized.
If you’ve found a solution, please share it here.
If you still need help, feel free to let us know.
